Logging

tekrajchhetri opened the pull request:
This is a FastAPI skeleton that currently implements logging features.

@djarecka  @aaronkanzer

aaronkanzer commented:
@tekrajchhetri Thanks for tagging me -- let me know if/when you'd like any feedback! (saw you were still pushing commits so didn't want to review anything yet)

tekrajchhetri commented:
@aaronkanzer There are no more changes to be made for the logging. So, I would say you can now review and give me feedback so that I could improve if any! And thanks in advance for your time to review

tekrajchhetri commented:
@aaronkanzer Yes, the goal is to build larger Python based API (or microservices application) for the work that I am doing. And since there will be multiple microservices components, I wanted to setup a skeleton so that I can re-use it.

Yes, you understood correctly. There will be both ML and Non-ML components. For ML, e.g., LLM, I would re-use this skeleton to create new one making changes as required. Therefore, I would say that this is mostly for non-ml, although most of the things shouldn't change, in my view.

Here's a very-high level initial architecture, which should provide you more context about the work.

<img width="865" alt="image" src="https://github.com/sensein/fastapi-skeleton/assets/52251022/a8d39929-6ee1-428c-9c99-753b7daac8b9">

aaronkanzer commented:
> @aaronkanzer Yes, the goal is to build larger Python based API (or microservices application) for the work that I am doing. And since there will be multiple microservices components, I wanted to setup a skeleton so that I can re-use it.
> 
> Yes, you understood correctly. There will be both ML and Non-ML components. For ML, e.g., LLM, I would re-use this skeleton to create new one making changes as required. Therefore, I would say that this is mostly for non-ml, although most of the things shouldn't change, in my view.
> 
> Here's a very-high level initial architecture, which should provide you more context about the work.
> 
> <img alt="image" width="865" src="https://private-user-images.githubusercontent.com/52251022/322857592-a8d39929-6ee1-428c-9c99-753b7daac8b9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyODYwNTQsIm5iZiI6MTcxMzI4NTc1NCwicGF0aCI6Ii81MjI1MTAyMi8zMjI4NTc1OTItYThkMzk5MjktNmVlMS00MjhjLTljOTktNzUzYjdkYWFjOGI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE2VDE2NDIzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1MDRhMWRiNzg1MWUxNWI4MTA4NmUyNzg4NTRmM2ZiZGQ3MzBjM2NmNWZlM2UwMzc4ZmE1Yzk2NmI1NGQ1ZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oZO3hmLSrwbjQfiTrGSXsorncgUnE7RPM3zPIAZ1EbE">

Thanks for this architecture diagram -- a few of thoughts/questions after reading your code:

• I'd re-evaluate how you go about authentication/permissions -- managed services like AWS Cognito will allow you to focus more on the actual science you are doing here (Knowledge Graph exploration with DANDI, etc.) instead of things like authentication which doesn't need to be custom-coded unless there is something special about your application. I think going the managed route will save you a lot of time and headache.

• I'd strongly suggest evaluating an architecture that leverages serverless technologies (look into AWS API Gateway + Lambdas + Cognito) -- LLM-related endpoints require a lot of bug-free code to not time out/obstruct other operations of your server if you are not in a serverless framework. [AWS Lamdbas have even implemented streaming support so you can provide quicker feedback](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/) in your UI for LLM-related activities that may take more time behind the scenes

• Do you have plans for a vectorized DB to best handle context for your LLM-related operations? ([something like Pinecone](https://www.pinecone.io/) is what I'm thinking. The reason I ask is so that you can disregard SQLAlchemy (and, by association, Alembic for your downstream ORM operations) and use something more simplistic and faster (such as DynamoDB)

Let me know if you have any further questions/thoughts -- happy to talk further.

(Cc @satra @hvgazula -- just tagging you here for context of some of the challenges that currently present themselves with packaging web apps around LLMs)

aaronkanzer commented on requirements.txt:
```diff
@@ -0,0 +1,22 @@
+fastapi==0.110.1
+uvicorn[standard]==0.29.0
+sqlalchemy
```
You pinned other versions, but not this -- any reason why?

aaronkanzer commented on core/routers/index.py:
```diff
@@ -0,0 +1,26 @@
+# @File    : index.py
+# @Software: PyCharm
+from fastapi import APIRouter
+from fastapi import Header
```
I don't think you are using this import here -- can probably remove

aaronkanzer commented on core/main.py:
```diff
@@ -0,0 +1,27 @@
+logger = logging.getLogger(__name__)
+
+app = FastAPI()
+configure_logging()
```
You can throw this call in one of the FastAPI lifecycle events upon startup

https://fastapi.tiangolo.com/advanced/events/#startup-event

tekrajchhetri replied:
Thanks @aaronkanzer. I will look into your comments and update the code accordingly.

tekrajchhetri commented:
@aaronkanzer I have updated the code as per your comments.

- configure_logging() will be triggered on startup
- removed unused import - Header
- sqlalchemy and other JWT related library have been removed as they're there for JWT based authentication.

aaronkanzer commented on core/databases.py:
```diff
@@ -0,0 +1,43 @@
+# @File    : databases.py
+# @Software: PyCharm
+
+import datetime
+
+import sqlalchemy as sa
+
+# https://docs.sqlalchemy.org/en/20/core/metadata.html
+metadata = sa.MetaData()
+
+# JWT User Table
+jwt_user_table = sa.Table(
+    "jwt_user",
+    metadata,
+    sa.Column("id", sa.Integer, primary_key=True),
+    sa.Column("email", sa.String, unique=True, nullable=False),
+    sa.Column("password", sa.String, nullable=False),
+    sa.Column(
+        "created_at", sa.DateTime, nullable=False, default=datetime.datetime.utcnow
+    ),
+    sa.Column(
+        "updated_at",
+        sa.DateTime,
+        nullable=False,
+        default=datetime.datetime.utcnow,
+        onupdate=datetime.datetime.utcnow,
+    ),
+)
```
You removed SQLAlchemy -- so I don't think this file is necessary at this time? Unless you had other intentions for it?

aaronkanzer commented on requirements.txt:
```diff
@@ -0,0 +1,15 @@
+fastapi==0.110.1
```
What commands are you running locally to install your requirements? Are you doing this via a local virtual environment? Docker container?

tekrajchhetri replied:
I will containerize the application but later will create a separate pull request. For now I am running in a local virtual environment.

aaronkanzer replied:
It seems you are using poetry as well (via the `pyproject.toml` file present in your PR) -- it might be useful to just use poetry to manage your dependencies as well here

tekrajchhetri replied:
Initially, I was using pyproject.toml for pycharm config but now I have moved all the dependencies to the poetry.

aaronkanzer commented:
@tekrajchhetri I think this PR and https://github.com/sensein/fastapi-skeleton/pull/2 can be merged! We can continue to iterate here as more of your application code is written
